README.md
# sfdc-qcp-vscode-extension README

This plugin improves the developer experience when working with [Salesforce QCP Quote Calculator Plugin](https://developer.salesforce.com/docs/atlas.en-us.cpq_dev_plugins.meta/cpq_dev_plugins/cpq_plugins_parent.htm) scripts.

Say goodbye to copy-and-paste, say hello to VSCode!

![commands](images/commands.png)

## Important Details

Locally, all files are saved as Typescript files, but there is no compilation so you will still need to write valid ES6 javascript instead.

The reason for this is because there are may times when someone else changes the remote file or a sandbox is refreshed, it would be nearly
impossible to pull the Javascript from SFDC and then turn that back into TypeScript.

In the future, it would be nice to have a "TypeScript mode" where the developers can truly be in bliss with TypeScript for development,
but that also means that the developer should never pull files down from SFDC, because Salesforce only knows about the JavaScript version of the file.

Your org information is encrypted with a unique key per workspace and is stored as an encrypted value in the `.qcp` directory and the extension creates or updates your `.gitignore` file to ensure this file does not get committed to source control. Even if it is committed to source control it will not work on other computers as the encryption key is stored locally on your computer related to the specific workspace.

## Features

This plugin comes with the following core features:

- Connect to a Salesforce Org
- Create example QCP files in your project as examples
- Pull all existing QCP records from Salesforce
- Pull specific QCP records from Salesforce
- Push all locals files to Salesforce and create or update existing records on Salesforce
- Push active file to Salesforce
- Push specific local file to Salesforce
- Backup local files to directory
- Backup all remote files to directory

### Available Commands

- `SFDC QCP: Validate org credentials`
  - Check if the current credentials are valid
- `SFDC QCP: Initialize Project`
  - Re-authorize or change org credentials
  - Pull all files from Salesforce (if new project)
  - Create example files and pull all (if new project)
- `SFDC QCP: Create example QCP files in your project`
  - Choose one or more examples files to add to your project (this overwrites any existing files with the same name)
- `SFDC QCP: Pull QCP records from Salesforce`
  - Pull all script files from Salesforce (this overwrites existing files with the same name)
- `SFDC QCP: Pull active file from Salesforce`
  - Refreshes the local open file with the version on Salesforce
  - This can also be triggered by clicking the download icon on the top right-hand side of the avtive editor
- `SFDC QCP: Push QCP files to Salesforce`
  - Choose one or more file sto push to Salesforce, which will create or update records in Salesforce
- `SFDC QCP: Push active file to Salesforce`
  - Push active file to Salesforce. If the file has unsaved changes, the file will first be saved
  - This can also be triggered by clicking the upload icon on the top right-hand side of the active editor
- `SFDC QCP: Backup local or remote files`
  - Local
    - This will copy all files from `/src` to a new folder called `/{date}`
  - Remote
    - This will fetch all files from Salesforce and copy to a new folder called `/{date}`
- `SFDC QCP: Compare records and show differences`
  - Compare a local file with a remote record
- `SFDC QCP: Compare active file with the Salesforce record`
  - Compare active local file with a remote record
  - This can also be triggered by clicking the external link icon on the top right-hand side of the active editor
- `SFDC QCP: View transpiled code for active file`
  - Opens the transpiled JavaScript from the active custom script record in Salesforce
- `SFDC QCP: View active QCP record in Salesforce`
  - Automatically logs in to Salesforce and navigated to the active custom script record detail page
  - This can also be triggered by clicking the external link icon on the top right-hand side of the active editor
- `SFDC QCP: Get QuoteModel record from Salesforce and save locally`
  - Shows a dialog to enter Quote Id and saves the QuoteModal as a JSON file in the data folder
  - This served two purposes:
    - Looking at the data for reference when building the QCP
    - Using the quoteModel and it's lineModels in unit tests

### Project Initialization

**To get started, create a blank folder and open it with VSCode.** You can use a folder with existing code if you want to.

Open the command pallette and type in `QCP` and choose `SFDC QCP: Initialize Project`. All other options will be hidden until the project is initialized.

When you initialize a new project, You will be redirected to Salesforce to login using OAuth, and be redirected back to VSCode.
Your authentication information will be stored in this file `.qcp/qcp-config.json`.

A `.gitignore` file will be created if one does not already exist and an entry for `.qcp` will be added to the file to ensure your credentials
are note submitted to source control.

The initialization will also create a `README.md` and `tsconfig.json` if they don't already exist.

Optionally, a `.prettierrc` file will be created during initialization. This can be configured in the extension settings.

Upon initialization, you will be asked if you want to pull all the files from Salesforce and optionally include example files locally.

## Requirements

- Salesforce Developer or Sandbox org.
- The Salesforce CPQ managed package must be installed installed.
- You must configure the CPQ Package level settings to choose which script to enable.

## Writing Unit Tests

When the project is initialized, a `/tests` folder will be created with a `qcp.spec.ts` file that can eb used for unit tests.
**Note**: if you created a project in older versions, you will not have the unit tests in your project. Simply run `SFDC QCP: Initialize Project` and re-authorize the same org and everything will get created again.

The file includes some example unit tests that are commented our, which you will need to uncomment and write the unit tests.

To get started with unit tests, perform the following steps:

1. Ensure you have at least one valid QCP file in the `/src` directory
2. Install the dependencies
   1. Install node if needed
   2. run `npm install` to install the developer dependencies referenced in the `package.json` file
3. Pull down some quote data that you would like to use for testing by running the command `SFDC QCP: Get QuoteModel record from Salesforce and save locally`
   1. Enter a valid quote ID and choose a file name
4. Uncomment the code at the bottom of the unit test file `/tests/qcp.spec.ts` and update the imports
   1. Update the filepath included in the example to point to your QCP file
   2. Replace the `quoteModel` import to point to the name of your downloaded quoteModel JSON file
5. If your QCP uses the `conn` [JSForce](https://jsforce.github.io/) object, then make sure you add valid credentials to the `.env` file

Here is an example of a working QCP and unit test

**/src/QCP.ts**

```typescript
export function onBeforeCalculate(quoteModel, quoteLineModels, conn) {
  return new Promise((resolve, reject) => {
    console.log('onBeforeCalculate()');
    getAccountCount(conn)
      .then(numAccounts => {
        console.log('numAccounts: ', numAccounts);
        quoteModel.record.Num_Accounts__c = numAccounts;
        resolve();
      })
      .catch(err => {
        reject(err);
      });
  });
}

function getAccountCount(conn) {
  return new Promise((resolve, reject) => {
    conn
      .query('SELECT count() FROM Account')
      .then(numAccountsResults => {
        resolve(numAccountsResults.totalSize);
      })
      .catch(err => {
        console.log('Error querying accounts', err);
        resolve();
      });
  });
}
```

**/data/a1j50000006gHK7AAM.json**
_This is a truncated example_

```json
{
  "record": {
    "attributes": {
      "type": "SBQQ__Quote__c",
      "url": "/services/data/v45.0/sobjects/SBQQ__Quote__c/a1j50000006gHK7AAM"
    },
    "Id": "a1j50000006gHK7AAM",
    "Num_Accounts__c": null,
    ...
  "lineItems": [{...}],
  "lineItemGroups": [],
  "isPartial": false,
  "hasMultiSegmentLines": false,
  "customerTotal": 185,
  "currencyDecimalScale": 2,
  "channelDiscountsOffList": false,
  "calculationRequired": false,
  "applyPartnerDiscountFirst": false,
  "applyAdditionalDiscountLast": false
}
```

**/tests/qcp.spec.ts**

```typescript
import { expect } from 'chai';
import * as quoteModel from '../data/a1j50000006gHK7AAM.json';
import * as qcp from '../src/QCP';
import { getConn } from './init-jsforce';

async function initJsforceConn() {
  return await getConn();
}

describe('QCP Test', () => {
  it('Should successfully call onBeforeCalculate()', async () => {
    const conn = await initJsforceConn();
    await qcp.onBeforeCalculate(quoteModel, quoteModel.lineItems, conn);
    const totalSize = (await conn.query('SELECT count() FROM Account')).totalSize;
    expect(quoteModel.record.Num_Accounts__c).to.equal(totalSize);
  });
});
```

## Extension Settings

- `sfdcQcp.pushOnSave` - When a file is saved, show prompt asking if file should be pushed to Salesforce
- `sfdcQcp.prettier` - Determines if a .prettierrc file will be created on project initialization
- `sfdcQcp.saveLog` - Determines if a log file should be saved each time a record from Salesforce is pushed or pulled
- `sfdcQcp.maxLogEntries` - Determines the maximum number of entries in the log file.

## Known Issues

[View issues on GitHub](https://github.com/paustint/sfdc-qcp-vscode-extension/issues).

## Release Notes

Check out my [Medium article](https://medium.com/@paustint/getting-started-with-the-salesforce-cpq-quote-calculator-plugin-vscode-extension-718306ff40d4) on using the extension.

See changes in Changelog.

## Contributing

Contributions of all kinds will be considered. https://github.com/paustint/sfdc-qcp-vscode-extension


CHANGELOG.md
# Change Log

[unreleased]

### 1.0.4

**7-13-2019**

- Fixed npm reported security vulnerability (lodash).
- Added script to ensure npm modules are https.

### 1.0.3

**6-10-2019**

- Fixed npm reported security vulnerability.

### 1.0.2

**5-17-2019**

- Improved build process to use webpack as recommended by VSCode.

### 1.0.1

**5-17-2019**

- The authorization URL no longer worked because it was double encoded. (#48)

### 1.0.0

**4-6-2019**

- Added support for fetching a quoteModel and saving locally and running unit tests. (#42)
- Orgs were often listed as invalid and required re-authentication, even though the token refresh was successful. (#44)
- Removed example file for async/await because this was not transpiled correctly with the QCP transpiler.
- Added button icons to the editor navigation bar for the following use-cases:
  - To pull or push the active file that appear in the top right section of the tab bar of an active file.
  - To compare the active QCP file with the record from Salesforce.
  - To view the active QCP file in Salesforce.
- Added various [Octicons](https://octicons.github.com/) to commands. So far, the only icons visible are in the active editor navigation menu bar.
- Removed a number of commands that were confusing and focused on commands that work with the active file, as that is the most common use-case
  - Removed Commands:
    - `SFDC QCP: Pull specific QCP file from SFDC`
    - `SFDC QCP: Pull remote record from Salesforce`
    - `SFDC QCP: Push all files to Salesforce`
    - `SFDC QCP: Open QCP record from Salesforce`

### 0.5.0

**02-08-2019**

- Added ability to push active file. (#35)
- Updated version of vscode dependency. (#34)
- Added ability to view a file in Salesforce (requires re-initializing org before this will work). (#35)
- Added ability to view the transpiled JavaScript from a record in Salesforce Salesforce. (#37)
- Added ability to pull active files and overwrite local version. This was previously available, but required specifically selecting the file to pull. (#38)

### 0.4.0

**01-18-2019**

- When credentials were updated, the in-memory version was not updated if prior credentials were valid (#25)
- Updated login process to use OAuth instead of username+password (#4)
  - User is redirected to login page in web browser, and is redirected back to the application
  - Credentials are then stored locally using an access and refresh token
  - When the token expired, a new token us automatically obtained
  - Added encryption to credentials to avoid storing in plaintext
    - The generated encryption key is unique for each workspace
- Added logging output to the output console to allow users to keep track of session history (#29)
- When a file is deleted locally, a prompt is shown to provide the option to also delete from Salesforce.
- Pushing files on Windows did not work properly as `c:\` was added twice to path. (#32)

### 0.3.0

**12-19-2018**

- Open sourced project.
- Added license.
- Added repository information.

### 0.2.4

**12-16-2018**

- Fixed typo with login text input.

### 0.2.3

**12-16-2018**

- Fixed type error in example `qcp-example-true-end-date-and-sub-term.ts`.
- Fixed bug with invalid credentials with viewing unsaved records from Salesforce.
- Initialize project is the only shown menu option for projects that do not have the qcp configuration file in the workspace.
- Published [blog article](https://medium.com/@paustint/getting-started-with-the-salesforce-cpq-quote-calculator-plugin-vscode-extension-718306ff40d4).

### 0.2.2

**12-15-2018**

- Modified extension icon.

### 0.2.1

**12-15-2018**

- Added extension icon.
- Updated theme.
- Updated extension description.
- If a local file is created and pushed, and a record in Salesforce with the same name exists, that record will be used and overwritten.
- Added command to view record in Salesforce without pulling the record data to a local file.

### 0.2.0

**12-15-2018**

- Updated changelog file. (#9)
- Misc code cleanup.
- Pushing files now allows selecting multiple files instead of just one or all. (#3)
- When pulling files, a prompt with various actions is presented to the user before overwriting local records. (#15)
- Added a log file in the `.qcp` directory to show a history of what was pushed and pulled. (#14)
- Added ability to push files when they are saved, which includes a user confirmation. (#13)
- Updated background on the Marketplace. (#10)
- Updated extension display name to `Salesforce CPQ - Quote Calculator Plugin`. (#10)
- On initialize, if an org is already configured, then there is an option to skip re-initializing the org and just re-create any config files.
- Added prettier configuration file creation with project initialization. (#11)
- Added command to get diff from files or records.
- Added settings:
  - `sfdcQcp.pushOnSave` - When a file is saved, show prompt asking if file should be pushed to Salesforce.
  - `sfdcQcp.saveLog` - Determines if a log file should be saved each time a record from Salesforce is pushed or pulled.
  - `sfdcQcp.maxLogEntries` - Determines the maximum number of entries in the log file.
  - `sfdcQcp.prettier` - Determines if a .prettierrc file will be created on project initialization.
  - `prettierConfig` - Default [prettier configuration](https://prettier.io/docs/en/configuration.html) object. You must edit this configuration in JSON mode.

### 0.1.1

**12-09-2018**

- Added additional information on getting started. (#6)
- Fixed typo in README. (#7)

### 0.1.0

**12-08-2018**

- Initial release of the Plugin.


src/common/constants.ts
import { InputBoxOptions, QuickPickItem, Uri } from 'vscode';
import { CustomScriptBase, CustomScriptFile, AuthHttp } from '../models';
import { parameterize } from './utils';
import { validateId } from './sfdc-utils';

/**
 * This file contains extension constants
 */

export const IV_LENGTH = 16; // For AES, this is always 16
export const OUTPUT_PANEL_NAME = 'SFDC QCP';
export const MEMONTO_KEYS = {
  ENC_KEY: 'ENC_KEY',
};

export const CUSTOM_SCRIPT_API_NAME = 'SBQQ__CustomScript__c';

export const REGEX = {
  ENDS_WITH_DASH_NUM: /-\d+$/,
  ANY_CHAR: /./gi,
  SRC_DIR: /\/src\//i,
};

// global settings keys
export const SETTINGS = {
  NAMESPACE: 'sfdcQcp',
  ENTRIES: {
    PUSH_ON_SAVE: 'pushOnSave',
    PRETTIER: 'prettier',
    SAVE_LOG: 'saveLog',
    MAX_LOG_ENTRIES: 'maxLogEntries',
    PRETTIER_CONFIG: 'prettierConfig',
  },
  DEFAULTS: {
    PUSH_ON_SAVE: false,
    PRETTIER: true,
    SAVE_LOG: true,
    MAX_LOG_ENTRIES: 150,
    PRETTIER_CONFIG: {
      printWidth: 140,
      parser: 'typescript',
      semi: true,
      tabWidth: 2,
      useTabs: false,
      singleQuote: true,
      trailingComma: 'all',
      bracketSpacing: true,
      arrowParens: 'avoid',
      insertPragma: false,
    },
  },
};

export const GITIGNORE_CONTENTS = `

# Added by VSCode Plugin - SFDC QCP
.qcp
.env

`;

export const QP = {
  INIT_ORG_CONFIRM: {
    YES: 'Yes - Re-initialize org.',
    NO: 'No - Use org that is currently configured.',
  },
  INIT_QCP_EXAMPLE: {
    EXAMPLE: 'Start with example QCP files.',
    PULL: 'Pull all QCP files from org.',
    EXAMPLE_AND_PULL: 'Pull all QCP files and create example QCP file.',
  },
  INIT_ORG_TYPE_QUICK_ITEM: {
    SANDBOX: 'Sandbox',
    DEV: 'Developer',
    PROD: 'Production',
    CUSTOM: 'Custom URL',
  },
  EXAMPLES: {
    ALL: 'All example files',
  },
  BACKUP_CHOOSE_SRC: {
    LOCAL: 'Local - Copy all files in src folder to backup directory.',
    REMOTE: 'Remote - Fetch all files from remote and put into backup directory.',
  },
  PUSH_ON_SAVE_CONFIRM: {
    YES: 'Yes - Push file to Salesforce.',
    NO: 'Cancel',
  },
  DELETE_REMOTE_ON_DELETE_CONFIRM: {
    YES: 'Yes - Also remove record from Salesforce.',
    NO: 'No - Keep record that exists on Salesforce.',
  },
  OVERWRITE_CONFIRM: {
    BACKUP: 'BACKUP - Backup local file before overwriting from Salesforce.',
    OVERWRITE: 'Overwrite - Overwrite local file with the code from Salesforce.',
    SKIP: 'Skip - Keep local file. (default)',
    BACKUP_ALL: 'Backup All - Backup all files that would be overwritten from Salesforce.',
    OVERWRITE_ALL: 'Overwrite All - Overwrite this file and any subsequent files.',
    SKIP_ALL: 'Skip All - Keep all local files if the content on Salesforce is different.',
    CANCEL: 'Cancel - Cancel operation. Any saved files will be kept, but all subsequent files will not be saved.',
  },
  COMPARE_CONFIRMATION: {
    LOCAL_WITH_REMOTE: 'Compare a local file with the same remote record.',
    LOCAL_WITH_ANY_REMOTE: 'Compare a local file with any remote record.',
    LOCAL_FILES: 'Compare a local file with another local file.',
    REMOTE_RECORDS: 'Compare a remote record with another remote record.',
  },
};

export const MESSAGES = {
  INIT: {
    NO_WORKSPACE: 'There is no open folder, the initialize command only works if a folder is opened.',
    SUCCESS: 'Your project has been successfully configured.',
    ORG_VALID: 'Your credential are valid.',
    ORG_INVALID: 'Your credential are invalid.',
    EXAMPLE_FILES_COPIED: `Successfully copied example files`,
  },
  PULL: {
    ALL_RECS_SUCCESS: (count: number = 0) => `Successfully downloaded and saved ${count} file${count === 1 ? '' : 's'} from Salesforce.`,
    PROGRESS_ALL: 'Downloading all QCP files from Salesforce.',
    PROGRESS_ONE: (file: string) => `Downloading ${file} from Salesforce.`,
    PROGRESS_REMOTE_LIST: `Getting list of scripts from Salesforce.`,
    MULTIPLE_REMOTE_RECORDS: (name: string) =>
      `There are multiple records on Salesforce named "${name}"` +
      ', you should rename or delete these duplicate records from Salesforce and locally.',
  },
  PUSH: {
    SUCCESS: (recordName: string) => `Successfully pushed record ${recordName} to Salesforce.`,
    SUCCESS_COUNT: (count: number) => `Successfully pushed ${count} records to Salesforce.`,
    ERROR: 'There was an error pushing this file to Salesforce.',
    ERROR_W_FILENAME: (filename: string) => `There was an error pushing ${filename} to Salesforce.`,
    ERROR_W_EX: (filename: string, message: string) => `There was an error pushing ${filename} to Salesforce. ${message}`,
    ERROR_FILE_UNTITLED: 'Please save your untitled file before pushing to Salesforce.',
    ERROR_NOTE_SRC_DIR: 'Only files in the src directory can be pushed to Salesforce.',
    ERROR_SAVING_FILE: 'There was an error saving your modified active file, please save your changes and try pushing to Salesforce again.',
    ERROR_NO_ACTIVE_FILE: 'There is no active file to push to Salesforce.',
    PROGRESS_ONE: 'Pushing file to Salesforce.',
    PROGRESS_ONE_W_FILENAME: (filename: string) => `Pushing ${filename} to Salesforce.`,
    PROGRESS_MULTI: 'Pushing files to Salesforce.',
  },
  DELETE: {
    DELETE_REMOTE_ON_DELETE_SUCCESS: (recordId: string) => `Record ${recordId} successfully deleted from Salesforce.`,
    DELETE_REMOTE_ON_DELETE_FAIL: (recordId: string) =>
      `There was an error attempting to delete record ${recordId} the record from Salesforce.`,
  },
  BACKUP: {
    IN_PROGRESS: (src: string) => `Backing up files from ${src}.`,
    SUCCESS: (src: string, folderName: string) => `Successfully backed up ${src} files to ${folderName}.`,
  },
  COMPARE: {
    SUCCESS: ``,
    REMOTE_RECORD_NOT_FOUND: (recordId: string) => `Could not find record on salesforce with Id ${recordId}.`,
  },
  FETCH: {
    IN_PROGRESS: (recordId: string) => `Fetching quoteModel for quote ${recordId}.`,
  },
};

type INPUT_OPTIONS = {
  INIT_ORG_CONFIRM: () => QuickPickItem[];
  INIT_USERNAME_INPUT: (currValue?: string) => InputBoxOptions;
  INIT_PASSWORD_INPUT: (currValue?: string) => InputBoxOptions;
  INIT_API_TOKEN_INPUT: (currValue?: string) => InputBoxOptions;
  INIT_ORG_TYPE_QUICK_ITEM: (currValue?: string) => QuickPickItem[];
  INIT_ORG_TYPE_CUSTOM_INPUT: (currValue?: string) => InputBoxOptions;
  INIT_QCP_EXAMPLE: () => QuickPickItem[];
  INIT_QCP_EXAMPLE_ALL: (files: string[]) => QuickPickItem[];
  PULL_ONE_SHOW_FILE_LIST: (files: CustomScriptFile[]) => QuickPickItem[];
  PULL_ONE_REMOTE_SHOW_FILE_LIST: (files: CustomScriptBase[]) => QuickPickItem[];
  PUSH_SHOW_FILE_LIST: (uris: Uri[]) => QuickPickItem[];
  PUSH_ON_SAVE_CONFIRM: (filename: string) => QuickPickItem[];
  DELETE_REMOTE_ON_DELETE_CONFIRM: (recordId: string, filename: string) => QuickPickItem[];
  BACKUP_CHOOSE_SRC: () => QuickPickItem[];
  OVERWRITE_CONFIRM: (filename: string) => QuickPickItem[];
  COMPARE_CONFIRMATION: () => QuickPickItem[];
  FETCH_RECORD_ID: () => InputBoxOptions;
  FETCH_RECORD_FILE_NAME: (currValue: string) => InputBoxOptions;
};

export const INPUT_OPTIONS: INPUT_OPTIONS = {
  INIT_ORG_CONFIRM: () => [{ label: QP.INIT_ORG_CONFIRM.YES }, { label: QP.INIT_ORG_CONFIRM.NO }],
  INIT_USERNAME_INPUT: (currValue?: string) => ({
    prompt: 'Enter your Salesforce Username',
    ignoreFocusOut: true,
    value: currValue,
  }),
  INIT_PASSWORD_INPUT: (currValue?: string) => ({
    prompt: 'Enter your Salesforce Password',
    password: true,
    ignoreFocusOut: true,
    value: currValue,
  }),
  INIT_API_TOKEN_INPUT: (currValue?: string) => ({
    prompt: 'Enter your API Token (Required if your IP address is not white listed)',
    password: false,
    ignoreFocusOut: true,
    value: currValue,
  }),
  INIT_ORG_TYPE_QUICK_ITEM: (currValue?: string) => [
    {
      label: QP.INIT_ORG_TYPE_QUICK_ITEM.SANDBOX,
      description: 'https://test.salesforce.com',
      picked: currValue === QP.INIT_ORG_TYPE_QUICK_ITEM.SANDBOX,
    },
    {
      label: QP.INIT_ORG_TYPE_QUICK_ITEM.DEV,
      description: 'https://login.salesforce.com',
      picked: currValue === QP.INIT_ORG_TYPE_QUICK_ITEM.DEV,
    },
    {
      label: QP.INIT_ORG_TYPE_QUICK_ITEM.PROD,
      description: 'https://login.salesforce.com',
      picked: currValue === QP.INIT_ORG_TYPE_QUICK_ITEM.PROD,
    },
    {
      label: QP.INIT_ORG_TYPE_QUICK_ITEM.CUSTOM,
      description: 'https://{domain}.my.salesforce.com',
      picked: currValue === QP.INIT_ORG_TYPE_QUICK_ITEM.CUSTOM,
    },
  ],
  INIT_ORG_TYPE_CUSTOM_INPUT: (currValue?: string) => {
    const value = currValue || 'https://domain.my.salesforce.com';
    return {
      ignoreFocusOut: true,
      placeHolder: '',
      prompt: 'Custom URL',
      value,
      valueSelection: [8, value.indexOf('.my') || 14],
    };
  },
  INIT_QCP_EXAMPLE: () => [
    { label: QP.INIT_QCP_EXAMPLE.EXAMPLE, picked: true, alwaysShow: true },
    { label: QP.INIT_QCP_EXAMPLE.PULL, picked: false, alwaysShow: true },
    { label: QP.INIT_QCP_EXAMPLE.EXAMPLE_AND_PULL, picked: true, alwaysShow: true },
  ],
  INIT_QCP_EXAMPLE_ALL: (files: string[]) => {
    const items: QuickPickItem[] = [{ label: QP.EXAMPLES.ALL, alwaysShow: true }];
    files.forEach(file => {
      items.push({ label: file, alwaysShow: true });
    });
    return items;
  },
  PULL_ONE_SHOW_FILE_LIST: (files: CustomScriptFile[]) => {
    return files.map(file => ({ label: file.fileName, picked: false }));
  },
  PULL_ONE_REMOTE_SHOW_FILE_LIST: (records: CustomScriptBase[]) => {
    return records.map(record => ({
      label: record.Name,
      description: record.Id,
      detail: `Last Modified by ${(record.LastModifiedBy || record.CreatedBy).Username} at ${record.LastModifiedDate.substring(0, 19)}`,
      picked: false,
    }));
  },
  PUSH_SHOW_FILE_LIST: (uris: Uri[]) => {
    return uris.map(uri => ({ label: uri.fsPath, picked: false }));
  },
  PUSH_ON_SAVE_CONFIRM: (filename: string) => [
    { label: QP.PUSH_ON_SAVE_CONFIRM.YES, detail: filename },
    { label: QP.PUSH_ON_SAVE_CONFIRM.NO },
  ],
  DELETE_REMOTE_ON_DELETE_CONFIRM: (recordId: string, filename: string) => [
    { label: QP.DELETE_REMOTE_ON_DELETE_CONFIRM.YES, description: recordId, detail: filename },
    { label: QP.DELETE_REMOTE_ON_DELETE_CONFIRM.NO },
  ],
  BACKUP_CHOOSE_SRC: () => [{ label: QP.BACKUP_CHOOSE_SRC.LOCAL, picked: true }, { label: QP.BACKUP_CHOOSE_SRC.REMOTE, picked: false }],
  OVERWRITE_CONFIRM: (filename: string) => [
    { label: QP.OVERWRITE_CONFIRM.BACKUP, detail: `Current File: ${filename}`, alwaysShow: true },
    { label: QP.OVERWRITE_CONFIRM.OVERWRITE, detail: `Current File: ${filename}`, alwaysShow: true },
    { label: QP.OVERWRITE_CONFIRM.SKIP, detail: `Current File: ${filename}`, alwaysShow: true },
    { label: QP.OVERWRITE_CONFIRM.BACKUP_ALL, detail: `Current File: ${filename}`, alwaysShow: true },
    { label: QP.OVERWRITE_CONFIRM.OVERWRITE_ALL, detail: `Current File: ${filename}`, alwaysShow: true },
    { label: QP.OVERWRITE_CONFIRM.SKIP_ALL, detail: `Current File: ${filename}`, alwaysShow: true },
    { label: QP.OVERWRITE_CONFIRM.CANCEL, detail: `Current File: ${filename}`, alwaysShow: true },
  ],
  COMPARE_CONFIRMATION: () => [
    { label: QP.COMPARE_CONFIRMATION.LOCAL_WITH_REMOTE },
    { label: QP.COMPARE_CONFIRMATION.LOCAL_WITH_ANY_REMOTE },
    { label: QP.COMPARE_CONFIRMATION.LOCAL_FILES },
    { label: QP.COMPARE_CONFIRMATION.REMOTE_RECORDS },
  ],
  FETCH_RECORD_ID: () => ({
    prompt: 'Enter a Quote Record Id to Fetch',
    password: false,
    ignoreFocusOut: true,
    validateInput: (value: string) => {
      if (validateId(value || '')) {
        return null;
      } else {
        return 'Enter a valid 15 or 18 digit id';
      }
    },
  }),
  FETCH_RECORD_FILE_NAME: (currValue: string) => ({
    prompt: 'What would you like the file to be named?',
    password: false,
    ignoreFocusOut: true,
    value: currValue,
    validateInput: (value: string) => {
      if (value.toLowerCase().endsWith('json')) {
        return null;
      } else {
        return 'The file extension must be json';
      }
    },
  }),
};

export const FILE_PATHS = {
  CONFIG: {
    target: '.qcp/qcp-config.json',
  },
  LOG: {
    target: '.qcp/qcp-log.json',
    backup: '.qcp/qcp-log.bak.json',
  },
  README: {
    source: 'README.md',
    target: 'README.md',
  },
  TSCONFIG: {
    source: 'tsconfig.json',
    target: 'tsconfig.json',
  },
  PACKAGE_JSON: {
    source: 'package.json',
    target: 'package.json',
  },
  PRETTIER: {
    source: 'prettierrc',
    target: '.prettierrc',
  },
  QCP: {
    source: 'src/qcp-example.ts',
    target: 'src/qcp-example.ts',
  },
  TESTS: {
    source: 'tests',
    target: 'tests',
  },
  ENV: {
    source: '.env',
    target: '.env',
  },
  SRC: 'src',
  DATA: {
    directory: 'data',
  },
};

const QUERY_FIELDS_BASE = `Id, Name`;
const QUERY_FIELDS_USER_FIELDS = `CreatedById, CreatedDate, LastModifiedById, LastModifiedDate, CreatedBy.Id, CreatedBy.Name, CreatedBy.Username, LastModifiedBy.Id, LastModifiedBy.Name, LastModifiedBy.Username`;
const QUERY_FIELDS_WO_CODE = `${QUERY_FIELDS_BASE}, ${QUERY_FIELDS_USER_FIELDS}`;
const QUERY_FIELDS_ALL = `${QUERY_FIELDS_WO_CODE}, SBQQ__Code__c, SBQQ__TranspiledCode__c, SBQQ__GroupFields__c, SBQQ__QuoteFields__c, SBQQ__QuoteLineFields__c`;

export const QUERIES = {
  ALL_WITHOUT_CODE: () => `SELECT ${QUERY_FIELDS_WO_CODE} FROM SBQQ__CustomScript__c`,
  ALL_RECS: () => `SELECT ${QUERY_FIELDS_ALL} FROM SBQQ__CustomScript__c`,
  BY_ID_RECS: (id: string) => `SELECT ${QUERY_FIELDS_ALL} FROM SBQQ__CustomScript__c WHERE Id = '${id}'`,
  BY_NAME_RECS: (name: string) => `SELECT ${QUERY_FIELDS_ALL} FROM SBQQ__CustomScript__c WHERE Name = '${name}'`,
  BY_NAME_RECS_NO_CODE: (name: string) => `SELECT ${QUERY_FIELDS_WO_CODE} FROM SBQQ__CustomScript__c WHERE Name = '${name}'`,
  BY_NAME_RECS_COUNT: (name: string) => `SELECT count() FROM SBQQ__CustomScript__c WHERE Name = '${name}'`,
};

export const client_id = '3MVG9KsVczVNcM8yH1pNeimwzaNciPgPq5lCmYI32we9ERWVHCx.vFaFRs9ejsGSHDoWyb8RGInzZJjAHJsQa';

/**
 * Note: we are not providing a client_secret, and JSForce does not support omitting this paramter
 * This is why the URL is being built manually
 */
export const AUTH_HTTP: AuthHttp = {
  getUserAgentAuth: {
    url: (domain: string) => {
      const params = parameterize({
        response_type: `token`,
        client_id,
        scope: `api refresh_token web`,
        redirect_uri: 'vscode://paustint.sfdc-qcp-vscode-extension/auth_callback',
        prompt: 'login',
      });
      return `${domain}/services/oauth2/authorize?${params}`;
    },
  },
};
